src/frontend/qt_sdl/LuaMain.cpp

```diff
@@ -0,0 +1,583 @@
+    {
+        printf("Force Stoping...");
+        luaThread->terminate();//if lua thread is stuck terminate the thread
+    }
+    luaThread->wait();
+    printf("Closed");
+    event->accept();
+
+}
+
+LuaConsole::LuaConsole(QWidget* parent)
+{
+    this->setParent(parent);
+}
+
+void LuaConsole::onGetText(QString string)
+{
+    this->appendPlainText(string);
+    QScrollBar* bar = verticalScrollBar();
+    bar->setValue(bar->maximum());
+}
+void LuaConsole::onClear()
+{
+    this->clear();
+}
+
+void LuaThread::luaClearConsole()
+{
+    emit signalClearConsole();
+}
+
+void LuaScript::LuaPrint(QString string)
+{
+    LuaDialog->console->onGetText(string);
+}
+
+//Used by "while true framestep" loops
+void LuaThread::luaYeild()
+{
+    LuaThread::exec();   
+}
+
+void LuaThread::luaPrint(QString string)
+{
+    emit signalPrint(string);
+}
+
+void LuaThread::luaStateSave(QString filename)
+{
+    emit signalStateSave(filename);
+}
+
+void LuaThread::luaStateLoad(QString filename)
+{
+    emit signalStateLoad(filename);
+}
+
+std::vector<LuaFunction*> LuaFunctionList; // List of all lua functions.
+lua_State* MainLuaState;//Currently only support one lua script at a time
+bool frameFlag = false;
+QWidget* LuaScript::panel = nullptr;
+
+LuaFunction::LuaFunction(luaFunctionPointer cfunctionPointer,const char* LuaFunctionName,std::vector<LuaFunction*>* container)
+{
+    this->cfunction=cfunctionPointer;
+    this->name=LuaFunctionName;
+    container->push_back(this);
+}
+
+void LuaConsoleDialog::onStartStop()
+{
+    if(luaThread==nullptr or !(luaThread->isRunning()))
+        return;
+    if (!luaThread->wait(5000))//5 seconds
+    {
+        luaThread->terminate();//if lua thread is stuck terminate the thread
+        console->onGetText(QString("Force Stopped"));
+    }
+    luaThread->wait();
+}
+
+void LuaThread::luaStop()
+{
+    flagRunning=false;
+    if (isRunning())
+        exit();
+}
+
+
+void LuaThread::luaTogglePause()
+{
+    flagPaused = !flagPaused;
+    exit();   
+}
+
+void LuaThread::run()
+{
+    MainLuaState=nullptr;
+    std::string fileName = LuaDialog->currentScript.fileName().toStdString();
+    std::string filedir = LuaDialog->currentScript.dir().path().toStdString();
+    lua_State* L = luaL_newstate();
+    luaL_openlibs(L);
+    for(LuaFunction* function : LuaFunctionList)
+        lua_register(L,function->name,function->cfunction);
+    std::filesystem::current_path(filedir.c_str());
+    if (luaL_dofile(L,&fileName[0])==LUA_OK)
+    {
+        MainLuaState=L;
+        flagRunning=true;
+    }
+    else
+    {
+        LuaPrint(lua_tostring(L,-1));
+        printf("Error: %s\n", lua_tostring(L, -1));
+        MainLuaState=nullptr;
+        flagRunning=false;
+    }  
+    //Lua Main loop
+    flagPaused = false;
+    while (flagRunning)
+    {
+        exec();
+    }
+    flagPaused = true;
+    emit signalPrint("Stopped");
+}
+
+//Is called every frame advance
+void LuaThread::luaUpdate()
```

Comment:
if I understand everything correctly this will cause a race condition as this function can run simultanously with the emulation of the next frame?